[venstarthermostat] more functions issue enhancement 10823

raveydavies opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<[Venstarthermostat] adding more functionality to match the local API>

<This Pull Request adds several enhancements to the Venstar Thermostat binding, as per enhancement issue 10823
https://github.com/openhab/openhab-addons/issues/10823
Functionality added copies the functionality of the local API and is:
Fan Mode
Fan State
Schedule Mode
Schedule Part
Time Stamp for Run Time
Run Time in heat 1 mode
Run Time in heat 2 mode
Run Time in cool 1 mode
Run Time in cool 2 mode
Run Time in Aux 1 mode
Run Time in Aux 2 mode
Run Time in Free Cool mode

README is updated

>

<The updates have been tested in the Eclipse openHAB environment, and also on my Raspberry Pi with OpenHAB, using my thermostat device. I have a very simple heating system (central heating boiler), so could not test some channels with real equipment, but I have checked all channels update properly according to commands and the thermostat changes accordingly.

>

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

digitaldan commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -338,23 +394,80 @@ private VenstarAwayMode getAwayMode() {
+
+    private ZonedDateTime getTimestampRuntime(VenstarRuntime runtime) {
+        ZonedDateTime z = ZonedDateTime.ofInstant(Instant.ofEpochSecond(runtime.getTimeStamp()),
+                ZoneId.systemDefault());
```
Actually it looks like this is reported in the user's timezone (or at least the time zone the device is set in) , so no need to convert to the system default time zone (or use GMT if you have to pass a timezone) .  But this is actually more tricky as the runtime schedules are more complicated, i'll post that in another comment.

raveydavies replied:
I was struggling with this due to lack of familiarity with timestamps, epoch, etc (I'm really an amateur....). What should I put as the function argument to Instant.ofEpochSecond() instead of "ZoneId.systemDefault()"? It's very confusing. I also noticed the time displayed in the channel was 2 hours ahead of my home time zone.

digitaldan replied:
So this still needs to be resolved , or the time is always going to be off.

So this is kinda tricky, but i think you need to offset this back to UTC and then create a ZoneDate time.  Dealing with time zones is always painful.   

So maybe something like....
```java
ZoneId  zoneId = ZoneId.systemDefault();
ZonedDateTime now = LocalDateTime.now().atZone(zoneId);
int diff = now.getOffset().getTotalSeconds();
ZonedDateTime z = ZonedDateTime.ofInstant(Instant.ofEpochSecond(runtime.getTimeStamp() -diff), zoneId);
```

```

digitaldan reviewed (changes requested):
Hi, thanks so much for these improvements!!!! I'm scratching my head wondering why we did not at least implement Fan mode and state, as those seem pretty useful, so i think your changes will be well received!

I added some feedback, mostly small stuff.

A note about the Runtime history.  It looks like the API is sending the last 7 days of runtime history.  The timestamp for the last day is always right now, which  makes sense as its the current runtime of each type  until right now from the last entry, once midnight happens, it becomes yesterday's values (so the timestamp remains the same and its now not the last entry in that array).

I could see just having this last entry (current runtime) could be useful for historic reasons and graphing (so putting values in grafana, rrd, etc...).  In this case i would update the documentation to make this more clear, as the timestamp is confusing why it's alway right now and that also the runtime values start at midnight.  It confused me, which is why i went looking a little deeper.

We could also add channels for each of the last 7 days, this would be great for users wanting to display this info on their OH UI.  If we could start over, adding channel groups might have been nice, but you can't mix them in now.  I would vote for this option if you want to keep the runtime functionality.  In this case your channels might look like 
```XML
<channel id="day0Heat1Runtime" typeId="heat1Runtime"/>
<channel id="day1Heat1Runtime" typeId="heat1Runtime"/>
<channel id="day2Heat1Runtime" typeId="heat1Runtime"/>
....
<channel id="day0Cool2Runtime" typeId="cool2Runtime"/>
<channel id="day1Cool2Runtime" typeId="cool2Runtime"/>
<channel id="day2Cool2Runtime" typeId="cool2Runtime"/>
...
```
and so on. WDYT?

raveydavies commented:
Thanks for the comments!! I will take a look at them in detail later today when I have some free time. About the runtimes....I was also wondering about including them all, it should be fairly easy to do. What's a channel group?

digitaldan commented:
> Thanks for the comments!! I will take a look at them in detail later today when I have some free time. About the runtimes....I was also wondering about including them all, it should be fairly easy to do. What's a channel group?

See https://www.openhab.org/docs/developer/bindings/thing-xml.html#channel-groups , it allows you to define a group of channel types once and then refer to it many times (once for each day in this case).  Unfortunately you either use ONLY channel groups OR individual channels, you can't mix them in a binding, so adding them would mean we break backwards compatibility for users.  I think its ok, just means a little more copy and pasting in the thing xml file for each day, but in your code you can loop through the days and dynamically refer to the channels like (warning untested):

```java
 private static final String CHANNEL_FMT = "day%d_%s";
          ....


            Collections.reverse(runtimes);
            for (int i = 0; i < 7; i++) {
                VenstarRuntime rt = null;
                if (i < runtimes.size()) {
                    rt = runtimes.get(i);
                }
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_TIMESTAMP_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DateTimeType(getTimestampRuntime(lastRuntime)));
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_HEAT1_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DecimalType(rt.getHeat1()));
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_HEAT2_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DecimalType(rt.getHeat2()));
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_COOL1_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DecimalType(rt.getCool1()));
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_COOL2_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DecimalType(rt.getCool2()));
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_AUX1_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DecimalType(rt.getAux1()));
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_AUX2_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DecimalType(rt.getAux2()));
                updateIfChanged(String.format(CHANNEL_FMT, i, CHANNEL_FC_RUNTIME),
                        rt == null ? UnDefType.UNDEF : new DecimalType(rt.getFC()));
            }
```
I'm assuming there may not always be 7 days (like when its reset or something) 
The channel constants would need to be changed to start with a capital letter of course to keep the camerlCase pattern.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/README.md:
```diff
@@ -54,6 +54,22 @@ After adding the Inbox item, enter the user name and password from the physical
+| scheduleModeRaw    | Number             | Current Schedule mode Raw (Read Only)| 0(Disabled) 1(Enabled)                         |
+| schedulePart       | String             | Current Schedule Part        |                                                        |
+| schedulePartRaw    | Number             | Schedule Part Raw (Read Only)|0(Morning) 1(Day) 2(Evening) 3 (Night) 255 (Inactive)   |
+| timestampRuntime   | DateTime           | Time Stamp of last RT update |Binding only looks at the last day runtime              |
```
OK I will update it. By the way, how does this formatting of a table in Markdown (in Eclipse) work? Is there a tutorial on it somewhere? I'm not familiar at all with this stuff, so I was kind of trying to fit RT into a column, probably not necessary :-).

digitaldan replied:
So put everything you want into the table, don't worry about formatting too much, then before you push your changes, use http://markdowntable.com/ to have it format your table with the correct spacing and such.  This is how i do it, others may have a different approach.  Eclipse has a very poor markdown editor and previewer.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +104,141 @@
+		<state readOnly="true" pattern="%d mins"/>
+	</channel-type>
+
+	<channel-type id="freecoolRuntime" advanced="false">
```
OK will update it.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -338,23 +394,80 @@ private VenstarAwayMode getAwayMode() {
+        return z;
+    }
+
+    private int getHeat1Runtime(VenstarRuntime runtime) {
```
I was copying the approach that was in the code already. Yes, it seems a bit silly unless they are public functions.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -338,23 +394,80 @@ private VenstarAwayMode getAwayMode() {
     }
 
-    private void updateSettings(VenstarAwayMode away) {
+    private VenstarFanMode getFanMode() {
```
Agree. I'll delete them, they are causing me a lot of unnecessary scrolling when I'm coding, it's driving me crazy.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntime.java:
```diff
@@ -0,0 +1,93 @@
+ * @author Matthew Davies - Class added to add more functionality to binding
+ */
+public class VenstarRuntime {
+    int ts;
```
OK. I will update it. Why is that? Coding convention because the number will eventually become very big?

digitaldan replied:
yes, this will overrun a `int` in something like 2038 , which while a long while off, its still better to have consistency in coding conventions other wise we get a mess of ints and longs all over the place.

raveydavies commented:
> Hi, thanks so much for these improvements!!!! I'm scratching my head wondering why we did not at least implement Fan mode and state, as those seem pretty useful, so i think your changes will be well received!
> 
> I added some feedback, mostly small stuff.
> 
> A note about the Runtime history. It looks like the API is sending the last 7 days of runtime history. The timestamp for the last day is always right now, which makes sense as its the current runtime of each type until right now from the last entry, once midnight happens, it becomes yesterday's values (so the timestamp remains the same and its now not the last entry in that array).
> 
> I could see just having this last entry (current runtime) could be useful for historic reasons and graphing (so putting values in grafana, rrd, etc...). In this case i would update the documentation to make this more clear, as the timestamp is confusing why it's alway right now and that also the runtime values start at midnight. It confused me, which is why i went looking a little deeper.
> 
> We could also add channels for each of the last 7 days, this would be great for users wanting to display this info on their OH UI. If we could start over, adding channel groups might have been nice, but you can't mix them in now. I would vote for this option if you want to keep the runtime functionality. In this case your channels might look like
> 
> ```
> <channel id="day0Heat1Runtime" typeId="heat1Runtime"/>
> <channel id="day1Heat1Runtime" typeId="heat1Runtime"/>
> <channel id="day2Heat1Runtime" typeId="heat1Runtime"/>
> ....
> <channel id="day0Cool2Runtime" typeId="cool2Runtime"/>
> <channel id="day1Cool2Runtime" typeId="cool2Runtime"/>
> <channel id="day2Cool2Runtime" typeId="cool2Runtime"/>
> ...
> ```
> 
> and so on. WDYT?

Yes I agree with this as well. Do you think the timestamps are needed? Or would it be better to just have channels labeled "Today Heat1 Runtime", "Yesterday Heat1 Runtime", " 2 days ago Heat1 Runtime"........?

digitaldan commented:
> Yes I agree with this as well. Do you think the timestamps are needed?

It couldn't hurt, I would just  mark the channel as "advanced" so it not shown by default in the UI

>Or would it be better to just have channels labeled "Today Heat1 Runtime", "Yesterday Heat1 Runtime", " 2 days ago Heat1 Runtime"........?

Yes, exactly !  The labels and Readme can more clearly state "Today, Yesterday, 2 Days ago, 3 Days ago,......" and so on.

raveydavies commented:
@digitaldan I'm working on the above changes. It will take me a couple of days. I'll commit the code when it's done and tested.

raveydavies commented:
@digitaldan did my last commit come through OK? Something strange happened at my end, but to me it looks OK.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -415,6 +474,51 @@ private void updateData() {
+                    CHANNEL_FC_RUNTIME_DAY6 };
+            int nRuntimes = runtimes.size();// check how many runtimes are available, might not be seven if equipment
+                                            // was reset
+            for (int i = 0; i < nRuntimes; i++) {
```
My mistake, you're right.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -415,6 +474,51 @@ private void updateData() {
+            String[] aux2Runtimes = { CHANNEL_AUX2_RUNTIME_DAY0, CHANNEL_AUX2_RUNTIME_DAY1, CHANNEL_AUX2_RUNTIME_DAY2,
+                    CHANNEL_AUX2_RUNTIME_DAY3, CHANNEL_AUX2_RUNTIME_DAY4, CHANNEL_AUX2_RUNTIME_DAY5,
+                    CHANNEL_AUX2_RUNTIME_DAY6 };
+            String[] fCRuntimes = { CHANNEL_FC_RUNTIME_DAY0, CHANNEL_FC_RUNTIME_DAY1, CHANNEL_FC_RUNTIME_DAY2,
```
I thought it should be fC like the individual variable names. Why not?

digitaldan replied:
b/c you clearly use fc as a single word here `CHANNEL_FC_RUNTIME_DAY0`  not `CHANNEL_F_C_RUNTIME_DAY0` , and i agree `FC` is netter then `F_C`

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -415,6 +474,51 @@ private void updateData() {
+
+            runtimeData = Objects.requireNonNull(gson.fromJson(response, VenstarRuntimeData.class));
+            List<VenstarRuntime> runtimes = runtimeData.getRuntimes();
+            String[] timestamps = { CHANNEL_TIMESTAMP_RUNTIME_DAY0, CHANNEL_TIMESTAMP_RUNTIME_DAY1,
```
I agree it's getting messy. I didn't like it when I programmed it, but I'm not sure how to resolve it (your CHANNEL_NAME+i won't work, will it?). I was actually wondering if it would be cleaner just to provide the runtime string, and let users do whatever they want with it (with some explanation of the String content in the documentation). I suspect anyone trying to do anything with this data will be quite "hardcore" anyway.

raveydavies replied:
Thanks!! Sounds good....I will update it. I guess I had somehow forgotten the variable values are strings.

raveydavies replied:
What do you think about just providing the runtime string as it comes directly from the thermostat, as a simplification? I suppose as I've already done a lot of code around this, it's not worth reversing :-)

digitaldan replied:
> What do you think about just providing the runtime string as it comes directly from the thermostat,

Not sure what you mean?

raveydavies replied:
I was just thinking it would be simpler to provide the raw json response from '.../query/runtimes' and let people do whatever they want with it with a JSON / Regex transformation operation. That would avoid all the extra channels and code we currently have (I don't like the fact the binding currently has 56 channels and 2 classes for this - it's messy :-O). A json probably wouldn't be too difficult to handle for people that are likely to use the data (it's not really core thermostat functionality). The thermostat also gives alarms in a similar format, which would again require multiple channels and classes unless a JSON response is used.

digitaldan replied:
No, returning raw JSON is not a great choice here.  I am fine with the extra channels and classes, you can mark them as advanced, and they won't show up by default in the UI, so it should not intimidate  users.  Returning in JSON will ensure very few people would ever use this.  Again, if you dynamically create the channel names as we discussed a few times, then its just the thing XML which is long, but there are definitely bindings with more in them, so not out of the ordinary .

raveydavies replied:
@digitaldan OK. I did the updates you suggested anyway. They are in my latest commit/push in this pull request.

raveydavies commented:
@digitaldan I've made updates according to the comments, I think I addressed everything. Thanks for the coaching.....it's very useful for a self-taught programmer who never needed to code to any kind of standard or go through any peer review before :-)

digitaldan commented:
>I've made updates according to the comments, I think I addressed everything. Thanks for the coaching.....it's very useful for a self-taught programmer who never needed to code to any kind of standard or go through any peer review before :-)

No worries, these are nice enhancements, and i'm happy to help.

raveydavies commented:
@digitaldan Hi - I updated as per your review. I think I have finally got the freeCool camelCase right and the same everywhere!! I also made all the Runtime channels advanced. Everything seems to be working correctly with my thermostat. Let me know if anything else is needed. Thanks again.

raveydavies commented:
@digitaldan just wondering if it's worth including the "alerts" provided by the API while we're at it. I think that's the only functionality which is not included now, apart from some data only applicable on commercial devices. WDYT?

digitaldan commented:
HI looks good!   There are some checkstyle warnings you can find in `target/code-analysis/report.html` that need to be fixed, but after that i'm good.   

> @digitaldan just wondering if it's worth including the "alerts" provided by the API while we're at it. I think that's the only functionality which is not included now, apart from some data only applicable on commercial devices. WDYT?

That sounds great, but let's open another PR for that and get this one merged in the meantime.

raveydavies commented:
I've fixed those. See my last commit. I previously also saw a warning about first author contribution but that's gone now. All should be fine with my last commit.

raveydavies commented:
Thanks for the approval! I suppose another approval is needed before merging?

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/README.md:
```diff
@@ -71,10 +112,22 @@ Thing venstarthermostat:colorTouchThermostat:001122334455 "Venstar Thermostat (G
 Number:Temperature Guest_HVAC_HeatSetpoint  "Heat Setpoint [%d °F]" {channel="venstarthermostat:colorTouchThermostat:001122334455:heatingSetpoint"}
 Number:Temperature Guest_HVAC_CoolSetpoint  "Cool Setpoint [%d °F]" {channel="venstarthermostat:colorTouchThermostat:001122334455:coolingSetpoint"}
-Number Guest_HVAC_Mode                      "Mode [%s]"             {channel="venstarthermostat:colorTouchThermostat:001122334455:systemMode"}
+String Guest_HVAC_Mode                      "Mode [%s]"             {channel="venstarthermostat:colorTouchThermostat:001122334455:systemMode"}
```
OK

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/README.md:
```diff
@@ -71,10 +112,22 @@ Thing venstarthermostat:colorTouchThermostat:001122334455 "Venstar Thermostat (G
-Number Guest_HVAC_State                     "State [%s]"            {channel="venstarthermostat:colorTouchThermostat:001122334455:systemState"}
-Number Guest_Away_Mode                      "Mode [%s]"             {channel="venstarthermostat:colorTouchThermostat:001122334455:awayMode"}
+String Guest_HVAC_State                     "State [%s]"            {channel="venstarthermostat:colorTouchThermostat:001122334455:systemState"}
+String Guest_Away_Mode                      "Mode [%s]"             {channel="venstarthermostat:colorTouchThermostat:001122334455:awayMode"}
```
OK

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/README.md:
```diff
@@ -71,10 +112,22 @@ Thing venstarthermostat:colorTouchThermostat:001122334455 "Venstar Thermostat (G
-Number Guest_Away_Mode                      "Mode [%s]"             {channel="venstarthermostat:colorTouchThermostat:001122334455:awayMode"}
+String Guest_HVAC_State                     "State [%s]"            {channel="venstarthermostat:colorTouchThermostat:001122334455:systemState"}
+String Guest_Away_Mode                      "Mode [%s]"             {channel="venstarthermostat:colorTouchThermostat:001122334455:awayMode"}
+String Guest_Fan_Mode                      "Mode [%s]"             {channel="venstarthermostat:colorTouchThermostat:001122334455:fanMode"}
```
OK

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/README.md:
```diff
@@ -88,6 +141,19 @@ sitemap demo label="Venstar Color Thermostat Demo"
     Text item=Guest_HVAC_State
+    Switch item=Guest_Fan_Mode mappings=[auto=Auto, on=On]
+    Text item=Guest_Fan_State
+    Switch item=Guest_Schedule_Mode=[enabled=Enabled,disabled=Disabled]
```
OK

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/README.md:
```diff
@@ -88,6 +141,19 @@ sitemap demo label="Venstar Color Thermostat Demo"
+    Text item=Guest_aux1RuntimeDay0
+    Text item=Guest_aux2RuntimeDay0
+    Text item=Guest_freeCoolRuntimeDay0
+    
```
It's already there - line 142

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanModeSerializer.java:
```diff
@@ -0,0 +1,35 @@
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link VenstarSystemModeSerializer} parses system mode values
```
OK

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanStateSerializer.java:
```diff
@@ -0,0 +1,35 @@
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link VenstarSystemModeSerializer} parses system mode values
```
OK

lolodomo replied:
No, you just do it on another class but not on this one

raveydavies replied:
OK fixed it now

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanStateSerializer.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarFanStateSerializer implements JsonDeserializer<VenstarFanState> {
```
Not sure if this is possible - it is creating an error as the over-ridden Interface JsonDeserializer seems to be @Nullable

raveydavies replied:
also...why specifically for this class but not others?

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/README.md:
```diff
@@ -41,20 +41,61 @@ After adding the Inbox item, enter the user name and password from the physical
+
+### Runtime data
+
+The Venstar thermostat provides data about how many minutes the system has been running in each of the different modes (heat1, heat2, cool1, cool2, aux1, aux2, free cool) every day for the last 7 days. A time stamp is provided with each run time data set which represents the end of each day. The binding reads the runtime data and time stamps and provides them all as separate channels.
```
OK. But why the CRs? And do you mean each sentence or each line?

lolodomo replied:
The recommendation is to have each sentence on one line, not several sentences on the same line. It helps review of changes.

raveydavies replied:
ok

raveydavies commented:
@lolodomo I think I addressed your comments. Please review again.

lolodomo commented:
> I think I addressed your comments. Please review again.

No, you just consider some of the comments.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanStateSerializer.java:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+
```
Please remove this blank line

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntime.java:
```diff
@@ -0,0 +1,93 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarRuntime {
```
Please add @NonNullByDefault

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntime.java:
```diff
@@ -0,0 +1,93 @@
+package org.openhab.binding.venstarthermostat.internal.dto;
+
+/**
+ * The {@link VenstarSensor} represents a sensor returned from the REST API.
```
Please fix the comment

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarInfoData.java:
```diff
@@ -25,20 +25,29 @@
     VenstarSystemState state;
     VenstarSystemMode mode;
     VenstarAwayMode away;
+    VenstarFanMode fan;
+    VenstarFanState fanstate;
+    VenstarScheduleMode schedule;
+    VenstarSchedulePart schedulepart;
```
private should be added I guess.
Please also fix the other members.

raveydavies replied:
Yes. OK.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntime.java:
```diff
@@ -0,0 +1,93 @@
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarRuntime {
+    long ts;
+    int heat1;
+    int heat2;
+    int cool1;
+    int cool2;
+    int aux1;
+    int aux2;
+    int fc;
```
Please add private

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntimeData.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+
+/**
+ * The {@link VenstarSensorData} represents sensor data returned from the REST API.
```
Please fix the comment

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntimeData.java:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarRuntimeData {
```
Please add @NonNullByDefault

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntimeData.java:
```diff
@@ -0,0 +1,41 @@
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarRuntimeData {
+    List<VenstarRuntime> runtimes;
```
Please add private

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarRuntimeData.java:
```diff
@@ -0,0 +1,41 @@
+    List<VenstarRuntime> runtimes;
+
+    public VenstarRuntimeData() {
+        super();
```
Also set runtimes to an empty list = List.Of()

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarScheduleMode.java:
```diff
@@ -0,0 +1,56 @@
+package org.openhab.binding.venstarthermostat.internal.dto;
+
+/**
+ * The {@link VenstarSystemMode} represents the value of the system mode returned
+ * from the REST API.
```
Please fix the comment

raveydavies replied:
OK (again :-( ). The joys of copy/pasting.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarScheduleMode.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public enum VenstarScheduleMode {
```
Please add @NonNullByDefault

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarScheduleModeSerializer.java:
```diff
@@ -0,0 +1,35 @@
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link VenstarSystemModeSerializer} parses system mode values
+ * from the REST API JSON.
```
Please fix the comment

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarScheduleModeSerializer.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarScheduleModeSerializer implements JsonDeserializer<VenstarScheduleMode> {
```
Please add @NonNullByDefault if possible

raveydavies replied:
This won't work - seem reason as the other ones, JsonDeserializer contains a @Nullable

lolodomo replied:
So forget.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarSchedulePart.java:
```diff
@@ -0,0 +1,59 @@
+package org.openhab.binding.venstarthermostat.internal.dto;
+
+/**
+ * The {@link VenstarSystemMode} represents the value of the system mode returned
+ * from the REST API.
```
Please fix the comment

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarSchedulePart.java:
```diff
@@ -0,0 +1,59 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public enum VenstarSchedulePart {
```
Please add @NonNullByDefault

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarSchedulePartSerializer.java:
```diff
@@ -0,0 +1,35 @@
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link VenstarSystemModeSerializer} parses system mode values
+ * from the REST API JSON.
```
Please fix the comment

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarSchedulePartSerializer.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarSchedulePartSerializer implements JsonDeserializer<VenstarSchedulePart> {
```
Please add @NonNullByDefault if possible

raveydavies replied:
Won't work - same reason as the other ones, JsonDeserializer has a @Nullable

lolodomo replied:
So forget

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -93,10 +108,10 @@
 
     private static final int TIMEOUT_SECONDS = 30;
     private static final int UPDATE_AFTER_COMMAND_SECONDS = 2;
-
```
This was not bad to separate constants with a blank line.

raveydavies replied:
Not sure I understand sorry.

lolodomo replied:
This is minor, you suppressed a blank line that made a better separation between the constants and the class members.
This was just better before your change.

raveydavies replied:
ok

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -186,7 +205,28 @@ public void handleCommand(ChannelUID channelUID, Command command) {
                 log.debug("Setting away mode to  {}", value);
                 setAwayMode(value);
                 updateIfChanged(CHANNEL_AWAY_MODE_RAW, new StringType(value.toString()));
+            } else if (channelUID.getId().equals(CHANNEL_FAN_MODE)) {
```
This is better to reverse and put the constant at left.

raveydavies replied:
I don't understand what you mean. This follows the format of the other similar blocks in the code.

lolodomo replied:
Ok, keep it consistent with the existing code.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -186,7 +205,28 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                log.debug("Setting fan mode to  {}", value);
+                setFanMode(value);
+                updateIfChanged(CHANNEL_FAN_MODE_RAW, new StringType(value.toString()));
+            } else if (channelUID.getId().equals(CHANNEL_SCHEDULE_MODE)) {
```
This is better to reverse and put the constant at left.

raveydavies replied:
I don't understand what you mean. This follows the format of the other similar blocks in the code.

lolodomo replied:
Ok, keep it as it is

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanMode.java:
```diff
@@ -0,0 +1,56 @@
+        return friendlyName;
+    }
+
+    public static VenstarFanMode fromInt(int mode) {
```
Please add `throws IllegalArgumentException`

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanState.java:
```diff
@@ -0,0 +1,56 @@
+        return friendlyName;
+    }
+
+    public static VenstarFanState fromInt(int state) {
```
Please add `throws IllegalArgumentException`

raveydavies replied:
OK, seeing as it does :-O

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarScheduleMode.java:
```diff
@@ -0,0 +1,56 @@
+        return friendlyName;
+    }
+
+    public static VenstarScheduleMode fromInt(int mode) {
```
Please add `throws IllegalArgumentException`

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarSchedulePart.java:
```diff
@@ -0,0 +1,59 @@
+        return friendlyName;
+    }
+
+    public static VenstarSchedulePart fromInt(int part) {
```
Please add `throws IllegalArgumentException`

raveydavies replied:
OK

raveydavies replied:
I also noticed several other classes were missing this. I corrected them all, and correspondingly inserted try/catch statements in the handler

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -186,7 +205,28 @@ public void handleCommand(ChannelUID channelUID, Command command) {
                 updateIfChanged(CHANNEL_AWAY_MODE_RAW, new StringType(value.toString()));
+            } else if (channelUID.getId().equals(CHANNEL_FAN_MODE)) {
+                VenstarFanMode value;
+                if (command instanceof StringType) {
+                    value = VenstarFanMode.valueOf(((StringType) command).toString().toUpperCase());
+                } else {
+                    value = VenstarFanMode.fromInt(((DecimalType) command).intValue());
+                }
```
You need to catch exception(s).

raveydavies replied:
OK. please check when I commit.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:
```diff
@@ -186,7 +205,28 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                updateIfChanged(CHANNEL_FAN_MODE_RAW, new StringType(value.toString()));
+            } else if (channelUID.getId().equals(CHANNEL_SCHEDULE_MODE)) {
+                VenstarScheduleMode value;
+                if (command instanceof StringType) {
+                    value = VenstarScheduleMode.valueOf(((StringType) command).toString().toUpperCase());
+                } else {
+                    value = VenstarScheduleMode.fromInt(((DecimalType) command).intValue());
+                }
```
You need to catch exception(s).

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -21,6 +21,63 @@
+			<channel id="cool1RuntimeDay4" typeId="cool1Runtime"/>
+			<channel id="cool1RuntimeDay5" typeId="cool1Runtime"/>
+			<channel id="cool1RuntimeDay6" typeId="cool1Runtime"/>
+			<channel id="cool2RuntimeDay0" typeId="cool2Runtime"/>
```
You forgot channels cool2RuntimeDay1 to cool2RuntimeDay6

raveydavies replied:
OK. Silly me.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+		<state readOnly="true"/>
+	</channel-type>
+
+	<channel-type id="fanState">
+		<item-type>String</item-type>
+		<label>Fan State</label>
+		<description>Current Fan State</description>
+		<state readOnly="true">
+			<options>
+				<option value="off">Off</option>
+				<option value="on">On</option>
+			</options>
+		</state>
+	</channel-type>
```
Why not rather use a switch type ?
Just a comment, not a request for change.

raveydavies replied:
No idea. I'm learning here, and I know practically nothing about xml. What would that look like?

lolodomo replied:
@digitaldan : is it ok for you like that ? I believe it should rather be a switch.

lolodomo replied:
I would propose something like:
```
	<channel-type id="fanState">
		<item-type>Switch</item-type>
		<label>Fan State</label>
		<description>Current Fan State</description>
		<state readOnly="true" />
	</channel-type>
```
But of course, it would also require few little changes in the thing handler.

raveydavies replied:
Ah sorry I misunderstood your comment. I'm sure there was a reason I chose a string. Let me think.

digitaldan replied:
Fan state does match a switch type quite clearly as its state is ON/OFF so yes i would agree with using a switch.

lolodomo replied:
@raveydavies : is it clear for you ? You will certainly have to update your handler (handle command and update state) but the change is easy.

raveydavies replied:
@lolodomo Yes it's clear, working on it now, sorry didn't have much time the last few days

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+		<state readOnly="true"/>
+	</channel-type>
+
+	<channel-type id="scheduleMode">
+		<item-type>String</item-type>
+		<label>Schedule Mode</label>
+		<description>Current Schedule Mode</description>
+		<state readOnly="false">
+			<options>
+				<option value="disabled">Disabled</option>
+				<option value="enabled">Enabled</option>
+			</options>
+		</state>
+	</channel-type>
```
Why not rather use a switch type ?
Just a comment, not a request for change.

raveydavies replied:
see previous comment

lolodomo replied:
@digitaldan : is it ok for you like that ? I believe it should rather be a switch.

digitaldan replied:
This would be better as a switch i think , but i'm ok either way.  Thermostats can be a little weird with toggles, FanMode for example is ON/Auto, but not off, so a switch here would actually be confusing as there's nothing to turn off (the fan always has to run if heating or cooling, but can be optionally set to run independently )

lolodomo replied:
So you can ignore my comment for this one.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+	</channel-type>
+
+	<channel-type id="heat1Runtime" advanced="true">
+		<item-type>Number:Dimensionless</item-type>
```
Why using `Number:Dimensionless` ?
You are providing a DecimalType when updating the channel, so `Number` looks appropriate.

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+	</channel-type>
+
+	<channel-type id="heat2Runtime" advanced="true">
+		<item-type>Number:Dimensionless</item-type>
```
Why using `Number:Dimensionless` ?
You are providing a DecimalType when updating the channel, so `Number` looks appropriate.

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+	</channel-type>
+
+	<channel-type id="cool1Runtime" advanced="true">
+		<item-type>Number:Dimensionless</item-type>
```
Why using `Number:Dimensionless` ?
You are providing a DecimalType when updating the channel, so `Number` looks appropriate.

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+	</channel-type>
+
+	<channel-type id="cool2Runtime" advanced="true">
+		<item-type>Number:Dimensionless</item-type>
```
Why using `Number:Dimensionless` ?
You are providing a DecimalType when updating the channel, so `Number` looks appropriate.

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+	</channel-type>
+
+	<channel-type id="aux1Runtime" advanced="true">
+		<item-type>Number:Dimensionless</item-type>
```
Why using `Number:Dimensionless` ?
You are providing a DecimalType when updating the channel, so `Number` looks appropriate.

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+	</channel-type>
+
+	<channel-type id="aux2Runtime" advanced="true">
+		<item-type>Number:Dimensionless</item-type>
```
Why using `Number:Dimensionless` ?
You are providing a DecimalType when updating the channel, so `Number` looks appropriate.

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+	</channel-type>
+
+	<channel-type id="freeCoolRuntime" advanced="true">
+		<item-type>Number:Dimensionless</item-type>
```
Why using `Number:Dimensionless` ?
You are providing a DecimalType when updating the channel, so `Number` looks appropriate.

raveydavies replied:
OK

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -89,6 +146,141 @@
+		<item-type>Number:Dimensionless</item-type>
+		<label>Heat1 Runtime Minutes</label>
+		<description>Run time in heat1 mode in minutes (Day 0 = TODAY, Day 6 = 6 days ago)</description>
+		<state readOnly="true" pattern="%d mins"/>
```
Suggestion: rather use "min" than "mins".
Same for others.

raveydavies replied:
yes, fair comment

lolodomo reviewed (changes requested):
Final review

raveydavies commented:
Oh ok. I missed the others. I'm on vacation now so will take a look in a weeks time.

lolodomo commented:
@raveydavies : no time to finish ?

raveydavies commented:
Not yet. Will get onto it this weekend.

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanMode.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public enum VenstarFanMode {
```
OK

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanModeSerializer.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public class VenstarFanModeSerializer implements JsonDeserializer<VenstarFanMode> {
```
There is a @Nullable in the JsonDeserializer, might be a nightmare to sort out.

lolodomo replied:
So forget

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanState.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Matthew Davies - Initial contribution
+ */
+public enum VenstarFanState {
```
OK

raveydavies commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanMode.java:
```diff
@@ -0,0 +1,56 @@
+package org.openhab.binding.venstarthermostat.internal.dto;
+
+/**
+ * The {@link VenstarSystemMode} represents the value of the system mode returned
```
OK

lolodomo replied:
Please, the comment is still wrong (system mode).

raveydavies replied:
OK, done now

raveydavies commented:
@lolodomo I think I addressed all your comments this time. Not sure what to do about the @NonNullByDefault for the serializer classes - the JsonDeserializer interface would need to be modified. Thanks for the detailed feedback, and for catching some of my really silly mistakes.

lolodomo commented on bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/dto/VenstarFanModeSerializer.java:
```diff
@@ -0,0 +1,36 @@
+    public VenstarFanMode deserialize(JsonElement element, Type arg1, JsonDeserializationContext arg2)
+            throws JsonParseException {
+        int key = element.getAsInt();
+        return VenstarFanMode.fromInt(key);
```
As `VenstarFanMode.fromInt` can throw an `IllegalArgumentException`, either you have to catch it here or to add `throws IllegalArgumentException` to the method and catch it at the higher level.
The best option could be to catch the exception in this method and to log something and return null. But then please check that the caller is handling a null value.
This comment is applicable to all the `fromInt` methods and so it has to be changed at different places.

raveydavies replied:
@lolodomo I'm a bit lost here. Honestly, this is at the limit of my programming abilities. Happy to learn though....

I understand there are 2 options.
1. Add try/catch to all methods calling xxxx.fromInt(), then make sure any calling functions can handle a null return value. I can't figure out how to do this because there is a GsonBuilder variable in the VenstarThermostatHandler class, which uses the VenstarFanModeSerializer class, and the VenstarFanModeSerializer class and its functions don't seem to get called anywhere else - so where am I supposed to be checking for capability to handle null? 
2. Throw the IllegalArgumentException from the function and catch it higher up the chain. In that case, is the code below sufficient or do I need to add something else?

Excuse my questions. It's not easy to follow and the original code before I started working on it (and which I copy/pasted a lot) already contained some of these problems.

`public class VenstarFanModeSerializer implements JsonDeserializer<VenstarFanMode> {
    @Override
    public VenstarFanMode deserialize(JsonElement element, Type arg1, JsonDeserializationContext arg2)
            throws JsonParseException,IllegalArgumentException {
        int key = element.getAsInt();
        return VenstarFanMode.fromInt(key);
    }
}`

lolodomo replied:
I believe the proper solution is to catch `IllegalArgumentException` and then throw a `JsonParseException`.
Then your call to gson.fromJson will certainlya utomatically throw a JsonSyntaxException in such a case.
So your code should look like this:
```
    @Override
    public VenstarFanMode deserialize(JsonElement element, Type arg1, JsonDeserializationContext arg2)
            throws JsonParseException {
        int key = element.getAsInt();
        try {
            return VenstarFanMode.fromInt(key);
        } catch (IllegalArgumentException e) {
            throw new JsonParseException(e);
        }
    }
```
Letr me know if this is still not clear.

lolodomo commented:
@raveydavies : this is almost ok now. Please fix the two minor comments you did not handle properly and consider the last one I added about handling `fromInt` exception. Regarding the kind oif channels to be added, I would like to have @digitaldan opinion.

raveydavies commented:
Oops. I missed those other calls to the fromInt method. I will correct them. What do you mean about the types of channels? What are you referring to?

lolodomo commented:
> What do you mean about the types of channels? What are you referring to?

I mean the kind of item attached to the channel. It can be a switch, a string, a number, ... You chose "String" while I think "Switch" would have been more appropriate. You just made a choice, this is not a bug.

raveydavies commented:
@lolodomo updated again as per your review. Change the FanState Type to a switch, and added Exception Handling in the Serializer classes. Hope it's OK now....it works with my thermostat.

lolodomo reviewed (approved):
LGTM

lolodomo commented:
Tank you @raveydavies

kaikreuzer commented:
@lolodomo One quick wish when squashing+merging PRs: Please clean up the commit message when squashing to only keep the relevant information and a single sign-off line - otherwise they look like [this](https://github.com/openhab/openhab-addons/commit/6c6c93e8f0c145ab9ac1cb59f4e8f6392e7afc4d). Thank you!

lolodomo commented:
@kaikreuzer : that is noted for next times.

kaikreuzer commented:
@lolodomo Many thanks. I know that this isn't documented anywhere and I should have mentioned this upfront, sorry for missing out on it.

raveydavies commented:
Thanks @lolodomo @digitaldan for all for the help and feedback. It was tough for me to get that right, but we got there 😅. I learned a lot again.

raveydavies commented:
@lolodomo I noticed a documentation error related to this PR. I forgot to change the item type for fan state to a Switch in the doc after changing it in the code. Is there a way to quickly update it or do I need to create another issue / PR?

lolodomo commented:
Yes  please create a new PR.
